include/openthread/dataset.h

```diff
@@ -113,11 +116,14 @@ typedef otIp6NetworkPrefix otMeshLocalPrefix;
  * This structure represents PSKc.
  *
  */
-OT_TOOL_PACKED_BEGIN
 struct otPskc
 {
-    uint8_t m8[OT_PSKC_MAX_SIZE]; ///< Byte values
-} OT_TOOL_PACKED_END;
+    union
+    {
+        uint32_t    m32;                    ///< Reference of the key.
+        uint8_t     m8[OT_PSKC_MAX_SIZE];   ///< Byte values
+    } mKeyMaterial;
+};
```

Comment:
One  question here on `otMasterKey` and `otPskc`.

These are public definitions intended for user of OT to use. Do we need the users of OT to be aware whether PSA is being used or not? 

I think it may be easier/better if we can hide this from user of OT API. Thoughts?

For example, when user gets/set the current active operational dataset (using `otDatasetGetActive()`) do we expect the user to get/set the `otPskc`/`otMasterKey` in the data as a literal `m8` byte array or a key ref?

Reply:
I think the user needs to set the `otPskc`/`otMasterKey` as literal arrays, but getting, we can hide the PSA operations from him. But how do we store the key refs then? One option is to just add a private variable to `MasterKey` and `Pskc` classes. But this would mean inheriting the `otMasterKey`/`otPskc` with no apparent use. Do you have any other suggestion for us to move forward if `otMasterKey`/`otPskc` are not made as unions?

But in any case we will need to stop the `otPskc` and `otMasterKey` from being saved as part of the dataset. It should only be stored in PSA ITS and we should not have a copy of this in the NVM. Any ideas on how we can do this?

Reply:
That is mainly my question, i.e. whether OT users need to be aware of PSA and manage key refs at next layer (what are the security implication of this).

I defeinitley see that it would be easier to not have user of API to be aware of involved in managing PSA and key refs and would favor this. If we go for this mode, then we can keep the public `otMasterKey` and `otPskc` definitions as before and define and use the `union` model types only in the core OT modules.

Regarding dataset we currently have two different public APIs:
1. `otDatasetGetActive()` which gives a parsed sruct of different dataset items (e.g. otMasterKey` and ...)
2. `otDatasetGetActiveTlvs()` which let user get the whole dataset as a sequence of TLVs.

In second case, the master key and PSKc are included in the TLVs (as literal sequence of bytes). If we keep the second API as is, then I see no advantage/reason of not providing literal key values from `otDatasetGetActive()` as well.

If we need to change the second API to exclude certain TLVs then we need more thought on how to do this.

---

> But in any case we will need to stop the otPskc and otMasterKey from being saved as part of the dataset. It should only be stored in PSA ITS and we should not have a copy of this in the NVM. Any ideas on how we can do this?

Do we need this in general? Do we need this as  part of this PR (adding PSA support)?

If we do, I don't have any immediate suggestion/ideas. I I think it requires some thoughts and designs to support such a behavior (does not seem to be a simple change to me).

Reply:
> That is mainly my question, i.e. whether OT users need to be aware of PSA and manage key refs at next layer (what are the security implication of this).

I think that OT users should have access to both APIs: To use key literals, or to use PSA key refs.

Most probably most users would like to use key literals, because that's easier.

However, there may be very secure application running OpenThread. OpenThread should work in Non-Secure Processing Environment. And in very secure applications, key literals shall never be present in NSPE. That's why in such application keys are retrieved from some source (could be out-of-band commissioning) and stored in Secure Processing Environment. OpenThread API user would have access only to key reference, but not to its value. To configure OpenThread, API user would need to pass key reference to OpenThread.

But for sure, we don't want to support this ultra-secure scenario in this PR. However, I think we should keep in mind that such scenario may arise and have public APIs prepared for this. One example of API preparation could be an extendable union containing only key literal at the moment, but could be extended with PSA key reference in the future.

Reply:
Yes, we really need a way to avoid storing keys in plaintext as part of dataset. And we will need another PR to address this issue.

Reply:
@abtink @hubertmis @Damian-Nordic @LuDuda I have finished all the updates requested and also solved the issues with commissioning. The changes work well at my desk but there are some issues with the running make pretty, also jobs are failing. I am fixing these as we speak, but in the mean time can I ask you guys to review the new changes?

Reply:
Also are C11 extensions allowed in OT?

Reply:
Looks like we don't support it. I will revert the unnamed unions back.